pkg/controller/endpointslice/endpointslice_controller.go

```diff
@@ -392,26 +393,18 @@ func (c *Controller) ensureSetupManagedByAnnotation(service *v1.Service) error {
 		return err
 	}
 
-	for _, endpointSlice := range endpointSlices {
-		if _, ok := endpointSlice.Labels[discovery.LabelManagedBy]; !ok {
-			if endpointSlice.Labels == nil {
-				endpointSlice.Labels = make(map[string]string)
-			}
-
-			endpointSlice.Labels[discovery.LabelManagedBy] = controllerName
-			_, err = c.client.DiscoveryV1beta1().EndpointSlices(endpointSlice.Namespace).Update(endpointSlice)
+	for _, epSlice := range endpointSlices {
+		if _, ok := epSlice.Labels[discovery.LabelManagedBy]; !ok {
+			epSlicePatch := []byte(fmt.Sprintf(`{"metadata":{"labels":{"%s":"%s"}}}`, discovery.LabelManagedBy, controllerName))
+			_, err = c.client.DiscoveryV1beta1().EndpointSlices(epSlice.Namespace).Patch(epSlice.Name, types.StrategicMergePatchType, epSlicePatch)
 			if err != nil {
 				return err
 			}
 		}
 	}
 
-	if service.Annotations == nil {
-		service.Annotations = make(map[string]string)
-	}
-
-	service.Annotations[managedBySetupAnnotation] = managedBySetupCompleteValue
-	_, err = c.client.CoreV1().Services(service.Namespace).Update(service)
+	svcPatch := []byte(fmt.Sprintf(`{"metadata":{"annotations":{"%s":"%s"}}}`, managedBySetupAnnotation, managedBySetupCompleteValue))
```

Comment:
same comment about using structured json construction and MergePatch (this patch can actually be made once and shared between invocations of the function)